store/stat_store.go

```diff
@@ -0,0 +1,442 @@
+	defer ss.RUnlock()
+
+	// Obtain latest tpBucket
+	lastElem := ss.buffer.Front()
+	if lastElem == nil {
+		return TimePoint{}, fmt.Errorf("the statStore is empty")
+	}
+
+	lastEntry := lastElem.Value.(tpBucket)
+
+	// create the latest TimePoint from lastEntry
+	lastTP := TimePoint{
+		Timestamp: ss.start.Add(time.Duration(ss.tpCount-1) * ss.resolution),
+		Value:     lastEntry.value,
+	}
+
+	return lastTP, nil
+}
+
+// fillCache caches the average, max and percentiles of the statStore.
+// assumes a write lock is taken by the caller.
+func (ss *statStore) fillCache() {
+
+	// calculate the average value
+	sum := uint64(0)
+	for elem := ss.buffer.Front(); elem != nil; elem = elem.Next() {
+		entry := elem.Value.(tpBucket)
+		sum += uint64(entry.count) * entry.value
+	}
+	ss.cachedAverage = sum / uint64(ss.tpCount)
+
+	// calculate the max value
+	curMax := uint64(0)
+	for elem := ss.buffer.Front(); elem != nil; elem = elem.Next() {
+		entry := elem.Value.(tpBucket)
+		if entry.max > curMax {
+			curMax = entry.max
+		}
+	}
+	ss.cachedMax = curMax
+
+	// sort all values in the statStore
+	vals := []float64{}
+	for elem := ss.buffer.Front(); elem != nil; elem = elem.Next() {
+		entry := elem.Value.(tpBucket)
+		for i := uint8(0); i < entry.count; i++ {
+			vals = append(vals, float64(entry.value))
+		}
+	}
+	sort.Float64s(vals)
+
+	// calculate all percentiles
+	ss.cachedPercentiles = []uint64{}
+	for _, spc := range ss.supportedPercentiles {
+		pcIdx := int(math.Trunc(spc * float64(ss.tpCount)))
+		ss.cachedPercentiles = append(ss.cachedPercentiles, uint64(vals[pcIdx]))
+	}
+
+	ss.validCache = true
+}
+
+// Average performs a weighted average across all buckets, using the count of -
+// resolutions at each bucket as the weight.
+func (ss *statStore) Average() (uint64, error) {
+	ss.RLock()
+	defer ss.RUnlock()
+
+	lastElem := ss.buffer.Front()
+	if lastElem == nil {
+		return uint64(0), fmt.Errorf("the statStore is empty")
+	}
+
+	if !ss.validCache {
+		ss.fillCache()
+	}
+	return ss.cachedAverage, nil
+}
+
+// Max returns the maximum element currently in the statStore.
+// Max does NOT consider the case where the maximum is in the last one minute.
+func (ss *statStore) Max() (uint64, error) {
+	ss.RLock()
+	defer ss.RUnlock()
+
+	if ss.buffer.Front() == nil {
+		return uint64(0), fmt.Errorf("the statStore is empty")
+	}
+
+	if !ss.validCache {
+		ss.fillCache()
+	}
+	return ss.cachedMax, nil
+}
+
+// Percentile returns the requested percentile from the statStore.
+func (ss *statStore) Percentile(p float64) (uint64, error) {
+	ss.RLock()
+	defer ss.RUnlock()
+
+	if ss.buffer.Front() == nil {
+		return uint64(0), fmt.Errorf("the statStore is empty")
+	}
+
+	// Check if the specific percentile is supported
+	found := false
+	idx := 0
+	for i, spc := range ss.supportedPercentiles {
+		if p == spc {
+			found = true
+			idx = i
+			break
+		}
+	}
+
+	if !found {
+		return uint64(0), fmt.Errorf("the requested percentile is not supported")
```

Comment:
I don't believe that could be the case. Percentiles are declared in the 0-1 range. If someone is requesting the 95.4853th percentile then it might be a problem but for two decimal digits this should be accurate